### PR TITLE
feat: [SSCA-298]: Added enforcement API in SSCA Manager

### PR DESCRIPTION
**Pull request by arpit-harness**
## Describe your changes
Added Enforcement API in SSCA Manager. This API recreates the SSCA Core enforce sbom API.

## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50639)

### CODE REVIEW

**Review by williamwissemann — DISMISSED** [review]

File: ssca-manager/config/ssca-manager-config.yml [hunk 1]
```diff
@@ -34,9 +34,16 @@ mongo:
...
 allowedOrigins:
   - "*"
 
 hostname: ${HOSTNAME:-localhost}
 
-basePathPrefix: ${BASE_PATH_PREFIX:-""}
\ No newline at end of file
+basePathPrefix: ${BASE_PATH_PREFIX:-""}
+
+ngManagerServiceSecret: IC04LYMBf1lDP5oeY4hupxd4HJhLmN6azUku3xEbeE3SUx5G3ZYzhbiwVtK4i7AmqyU9OZkwB4v8E9qM
```
Comment (williamwissemann) [inline ssca-manager/config/ssca-manager-config.yml:49]: This is probably going to get flagged by gitleaks

**Review by deepharness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 43 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50639)* status: Some of the checks are too old. 42 file(s) to review. 1 change requests. 

___
*[`ssca-manager/service/src/main/java/io/harness/ssca/beans/AllowLicense.java` line 25 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6HAEu--yd3HYm8VtZ:-Na6HAEv7XIDAY-uFa3u:b37sg4w) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/beans/AllowLicense.java#L25)):*
> ```Java
> 
>   public Document uniquePackageFilter() {
>     Document nameFilter = new Document("packagename", name);
> ```

use constants for normalized field names . You can get it from DTO
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/beans/AllowLicense.java` line 31 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6HdVC-AqEsf1BBU6Q:-Na6HdVC-AqEsf1BBU6R:b37sg4w) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/beans/AllowLicense.java#L31)):*
> ```Java
>       regexMap.put("$regex", deniedVendors);
>       regexMap.put("$option", "i");
>       Document filter = new Document("packagelicense", new Document(regexMap));
> ```

use constants for normalized field names . You can get it from DTO
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/beans/AllowLicense.java` line 37 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6HeBnDgaS9fyjbmQH:-Na6HeBnDgaS9fyjbmQI:b37sg4w) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/beans/AllowLicense.java#L37)):*
> ```Java
>       regexMap.put("$regex", license);
>       regexMap.put("$option", "i");
>       Document filter = new Document("packagelicense", new Document(regexMap));
> ```

use constants for normalized field names . You can get it from DTO
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/beans/AllowList.java` line 25 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6J73KDpOEnkkClcDP:-Na6J73KDpOEnkkClcDQ:bmtqqhy) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/beans/AllowList.java#L25)):*
> ```Java
>   AllowListItem allowListItem;
>   @Override
>   public PolicyType getType() {
> ```

We should use generics for EnfocementLIst<AllowList> or EnfocementLIst<DenyList>
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/beans/DenyList.java` line 29 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6JMRZ5bXKz1Hfv0S8:-Na6JMRZ5bXKz1Hfv0S9:bmtqqhy) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/beans/DenyList.java#L29)):*
> ```Java
>   @Data
>   @Builder
>   public static class DenyListItem implements EnforcementListItem {
> ```

We should use generics for EnfocementLIst<AllowList> or EnfocementLIst<DenyList>
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/beans/Supplier.java` line 25 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6I5g93MFW3ZHInZN8:-Na6I5g93MFW3ZHInZN9:b37sg4w) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/beans/Supplier.java#L25)):*
> ```Java
> 
>   public Document uniquePackageFilter() {
>     Document nameFilter = new Document("packagename", name);
> ```

use constants for normalized field names . You can get it from DTO
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/beans/Supplier.java` line 31 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6I6jBA9gn4vGLqom6:-Na6I6jBA9gn4vGLqom7:b37sg4w) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/beans/Supplier.java#L31)):*
> ```Java
>       regexMap.put("$regex", deniedVendors);
>       regexMap.put("$option", "i");
>       Document filter = new Document("packageoriginatorname", new Document(regexMap));
> ```

use constants for normalized field names . You can get it from DTO
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/enforcement/mongo/MongoExecutor.java` line 30 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6JZP58qt35U_plSSJ:-Na6JZP58qt35U_plSSK:b6qpdjq) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/enforcement/mongo/MongoExecutor.java#L30)):*
> ```Java
> 
> @Slf4j
> public class MongoExecutor implements IRuleExecutor {
> ```

Should use generics its breaking SRP. There should be 2 implementation AllowListMongoExecutor & DenyListMongoExecutor
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/enforcement/mongo/MongoExecutor.java` line 38 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6KPko17-8mFgdTik4:-Na6KPko17-8mFgdTik5:b-gampuc) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/enforcement/mongo/MongoExecutor.java#L38)):*
> ```Java
>   @Override
>   public List<EnforcementResultEntity> execute(Engine engine) {
>     switch (engine.getRules().getType()) {
> ```

No switch statement needed if we use generics
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/enforcement/rule/IRuleInterpreter.java` line 15 at r1](https://reviewable.io/reviews/harness/harness-core/50639#-Na6KtthBCLX2qVy3kXb:-Na6KtthBCLX2qVy3kXc:b-h91009) ([raw file](https://github.com/harness/harness-core/blob/e10d9f86925f0e341aef28d56dd48b4640e050d9/ssca-manager/service/src/main/java/io/harness/ssca/enforcement/rule/IRuleInterpreter.java#L15)):*
> ```Java
> import org.springframework.data.mongodb.core.query.Query;
> 
> public interface IRuleInterpreter {
> ```

Should use generics interpretRules<T>(EnforcementListItem item, String orchestrationId, T type); where T is AllowListRule or DenyListRule

**Review by deepharness — CHANGES_REQUESTED** [review]
> Reviewed 6 of 43 files at r1, 8 of 33 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50639)* status: 37 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/EnforcementService.java` line 18 at r2](https://reviewable.io/reviews/harness/harness-core/50639#-NaGvdMUEocBom92vm0X:-NaGvdMUEocBom92vm0Y:b9p04yj) ([raw file](https://github.com/harness/harness-core/blob/792a071094d6a518eff0b94e7dc513444bb8f429/ssca-manager/service/src/main/java/io/harness/ssca/services/EnforcementService.java#L18)):*
> ```Java
> import java.util.List;
> 
> public interface EnforcementService {
> ```

Prefer more specific name in future , as service is too generic & in future anyone can use this interface to put any method.
https://medium.datadriveninvestor.com/the-true-meaning-of-service-and-manager-class-names-d09a08731fd9
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/SbomProcessorServiceImpl.java` line 85 at r2](https://reviewable.io/reviews/harness/harness-core/50639#-NaGwrlS1S4zZX4dUptB:-NaGwrlS1S4zZX4dUptC:b-tnj5we) ([raw file](https://github.com/harness/harness-core/blob/792a071094d6a518eff0b94e7dc513444bb8f429/ssca-manager/service/src/main/java/io/harness/ssca/services/SbomProcessorServiceImpl.java#L85)):*
> ```Java
> 
>   @Override
>   public EnforceSbomResponseBody enforceSbom(
> ```

We should break down this class as its the single entry point for ssca domain. We should split it into 2 ProcessSBOMWorkflow & EnforceSBOMWorkflow
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/enforcement/executors/mongo/filter/allowlist/SupplierQueryBuilder.java` line 57 at r2](https://reviewable.io/reviews/harness/harness-core/50639#-NaGtm24EyuevU_JOcBU:-NaGtm24EyuevU_JOcBV:b-uyqmys) ([raw file](https://github.com/harness/harness-core/blob/792a071094d6a518eff0b94e7dc513444bb8f429/ssca-manager/service/src/main/java/io/harness/ssca/enforcement/executors/mongo/filter/allowlist/SupplierQueryBuilder.java#L57)):*
> ```Java
>     for (Supplier supplier : suppliers) {
>       filters.add(new Document(NormalizedSBOMEntityKeys.packageOriginatorName.toLowerCase(),
>           new Document("$not", new Document("$regex", supplier.getSupplier()))));
> ```

Use a enum or constants for all the mongo operators like $nin,$regex etc. we used across the all the classes. it will help later in quick overview & performance tuning
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/enforcement/executors/mongo/filter/denylist/fields/PurlField.java` line 31 at r2](https://reviewable.io/reviews/harness/harness-core/50639#-NaGtXhK3_eO1kQdrChG:-NaGtXhK3_eO1kQdrChH:b-eoa9jp) ([raw file](https://github.com/harness/harness-core/blob/792a071094d6a518eff0b94e7dc513444bb8f429/ssca-manager/service/src/main/java/io/harness/ssca/enforcement/executors/mongo/filter/denylist/fields/PurlField.java#L31)):*
> ```Java
>     List<String> splitPurl = Arrays.asList(denyListItem.getPurl().split("/"));
>     String deniedPackageManager = splitPurl.get(0).split(":")[1];
> ```

Put a comment on how the purl looks like here , it will help in understanding the switch cases

**Comment by arpit-harness** [comment]
trigger uts

**Review by deepharness — APPROVED** [review]
> Reviewed 6 of 43 files at r1, 11 of 33 files at r2, 4 of 13 files at r3, 24 of 24 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50639)* status: :shipit: complete! ✓

**Comment by arpit-harness** [comment]
trigger uts

**Comment by arpit-harness** [comment]
trigger ut7

**Pull request merged**